File: 卡码/代码随想录/1/06-二叉树/C++/02-二叉树属性/LeetCode_0113.cpp
//113. 路径总和 II
#include <iostream>
#include <vector>
#include "../TreeNode.h"
using namespace std;

class Solution {
	vector<vector<int>> res;
	vector<int> path;
public:
  vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
		traversal(root,targetSum);
		return res;
  }
  void traversal(TreeNode* root,int targetSum){
		if(!root) return;
		path.push_back(root->val);
		targetSum-=root->val;
		//叶子节点且sum减到0说明找到路径
		if(!root->left&&!root->right&&targetSum==0) res.push_back(path);
		traversal(root->left,targetSum);
		traversal(root->right,targetSum);
		path.pop_back();
	}
};

int main(){
  system("pause");
  return 0;
}

File: 卡码/代码随想录/1/06-二叉树/Python/02-二叉树属性/LeetCode_0113.py
#113. 路径总和 II
from typing import *
from ..TreeNode import *

class Solution:
  def __init__(self):
    self.res=[]
    self.path=[]
  def traversal(self,root:Optional[TreeNode],targetSum:int):
    if not root:
      return
    self.path.append(root.val)
    #叶子节点且sum等于当前节点值说明找到路径
    if not root.left and not root.right and targetSum==root.val:
      self.res.append(list(self.path))
    self.traversal(root.left,targetSum-root.val)
    self.traversal(root.right,targetSum-root.val)
    self.path.pop()
  def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
    self.traversal(root,targetSum)
    return self.res
